test/router.js

```diff
@@ -742,12 +742,12 @@ describe('Router', function () {
       .expect(404, cb)
 
       request(server)
-      .post('/foo')
+      .post('/foo/')
       .expect(200, 'saw POST /', cb)
 
       request(server)
       .post('/foo/bar')
-      .expect(200, 'saw POST /bar', cb)
+      .expect(404, cb)
```

Comment:
I mean, there is no way this is actually the intended behavior, right? If `/foo/` is not a prefix of `/foo/bar` I don't know what `.use` is supposed to be doing.

Reply:
Well you you be required to do `.use('/foo')` If you wanted to match `/foo/bar`.

Reply:
Yea, but then what will it match? If having the trailing / is not correct, then why not throw? I'm afraid this is going to be very confusing.

Reply:
Using https://wesleytodd.github.io/express-route-tester/ I choose 2.1.0 and it says `/foo/` with no check boxes matches `/foo/bar`. It seems either that test is wrong or there is something else going on here that needs to be fixed.

Reply:
`use('/foo/'` will match `/foo/` and nothing else.  It is not that it is "not correct", it just has a very specific use case.  If you wanted the old behavior, all you have to do is change to `use('/foo'` I think.

Reply:
You need to have `end` checked to match what is un a use call I think:

```
var layer = new Layer(path, {
  sensitive: this.caseSensitive,
  strict: false,
  end: false
}, fn)
```

Reply:
If end is checked, that means `end: true`, but like you posted, `end` is false, which is why  unchecked it. The `.use` is the in-ending paths.

Reply:
Yeah sorry, I should have stopped before posting that, deleted right after posting it lol.

Reply:
haha. Cleaned up the convo :)

Reply:
That defeats the entire purpose of `.use`, does it not? How is `/foo/` not a prefix of `/foo/bar` ? Does `/foo/bar` not begin with `/foo/` ?

Reply:
The point of `.use` is maintained without the trailing `/`, and I thought this was reasonable.  To get that behavior back we would have to do make an exception for the behavior of `path-to-regexp`.  AFAIK we only do that for `*` and `/`, adding this exception would be a bit more involved because it is any path with a trailing `/`, and possibly involve other edge cases.

Would it be alright if we directly documented this new behavior?  Or do you think working on this code to add this exception back at the router layer would be good?

Reply:
I'm not sure. To me, `.use('/foo/', fn)` seems to cause completely unexpected behavior from how Express.js `.use` has always worked forever before. It's in many books and I think breaking this is going to be a large uphill battle from a user support perspective.

I think that either (a) we need to make it work as expected or (b) it shouldn't even be an accepted argument and throw / error out early.

I'm actually curious, perhaps before we proceed, that if you were to document this "new" behavior, what would it sound like / what would the use-case for this type of argument for `.use` be? It just seems really weird if `.use` works two completely different ways depending just on if you happen to add a trailing `/` on the string argument. Seems very unintuitive, but maybe I just don't understand the use-case this new behavior is enabling?

Reply:
Also, perhaps the answer is that `path-to-regexp` no longer meets the needs of Express.js and we need to scrap it for something else. Just putting things on the table.

Reply:
> large uphill battle from a user support perspective

Yeah we should avoid that if reasonable.  FWIW, if I ever wanted to serve a path like this I never put a trailing `/`.  I am not sure how common either one is but every project I have worked on has not used the trailing `/`.



>  (b) it shouldn't even be an accepted argument and throw / error out early.

I think we need some version of the trailing slash if someone wants specifically to do something like serve a directory listing in this case.  So erroring is not really an option.

> I think that either (a) we need to make it work as expected

I am looking at what this would take.  I think it is more simple than I first thought.  See the new commit I added.  Is this satisfactory?

> Also, perhaps the answer is that path-to-regexp no longer meets the needs of Express.js

I would hope we can avoid this unless we actually wanted to change the routing/matching semantics significantly.